File: example/1d-linear-convection/weno3/julia/01f/python/reconstructor/factory.py
# reconstructor/factory.py
from .eno import EnoReconstructor
from .weno3 import Weno3Reconstructor
from cfd_registry import ComponentRegistry

class ReconstructorFactory:
    @staticmethod
    def create(config, domain):
        scheme = config.recon_scheme.lower()
        order = getattr(config, 'spatial_order', None)
        
        if scheme == "weno":
            if order is None:
                raise ValueError("使用 'weno' 时必须设置 config.spatial_order")
            scheme = f"weno{order}"
        
        # 使用BaseFactory，但处理特殊参数
        from factories.base_factory import BaseFactory
        
        try:
            if scheme == "eno":
                if order is None:
                    order = 3
                # ENO需要特殊参数
                return ComponentRegistry.create('reconstructor', scheme, order, domain.ntcells)
            elif scheme == "weno3":
                # WENO3无参数
                return BaseFactory.create_component('reconstructor', scheme)
            else:
                # 其他情况
                return BaseFactory.create_component('reconstructor', scheme, config)
        except ValueError as e:
            # 简单的回退逻辑
            if scheme == "eno":
                if order is None:
                    order = 3
                from .eno import EnoReconstructor
                return EnoReconstructor(order, domain.ntcells)
            elif scheme == "weno3":
                from .weno3 import Weno3Reconstructor
                return Weno3Reconstructor()
            raise

File: example/1d-linear-convection/weno3/julia/02g/reconstructor/factory.jl
# julia/reconstructor/factory.jl

"""
ReconstructorFactory
对标 Python: ReconstructorFactory.create(config, domain)
"""
module ReconstructorFactory

# ✅ 不要用 using ..EnoReconstructor（它们不是模块）
# 直接通过 Main. 引用顶层定义的类型

function create(config::Any, domain::Any)
    scheme = lowercase(string(getfield_safe(config, :recon_scheme, "")))
    if scheme == ""
        error("必须先通过 with_reconstruction 设置重建格式！")
    end

    # 处理 WENO 默认命名（如 Python）
    if scheme == "weno"
        order = getfield_safe(config, :spatial_order, 5)
        scheme = "weno$(order)"
    end

    if scheme == "eno"
        order = getfield_safe(config, :spatial_order, 3)
        return Main.EnoReconstructor(order, domain.ntcells)
    elseif scheme == "weno3"
        return Main.Weno3Reconstructor()
    else
        error("不支持的重建格式: $scheme（仅支持 eno/weno3）")
    end
end

# 辅助函数（复制自 domain.jl）
function getfield_safe(obj, name::Symbol, default)
    return isdefined(obj, name) ? getfield(obj, name) : default
end

end  # module ReconstructorFactory

export ReconstructorFactory